Components: Update PostSelector to render infinitely, using ReactVirtualized

aduth opened the pull request:
Fixes #3881
Related: #5135

This pull request seeks to reimplement `<PostSelector />` as an infinitely rendered list. This is the first step toward better management of persisted post query data (including proper persistence of found items, see a5f3529b82dd8862afdba2c5e6eff332996e3e58), as described at https://github.com/Automattic/wp-calypso/pull/5135#issuecomment-221672980 .

__Implementation notes:__

These changes include a new dependency, `react-virtualized`. There are a few reasons I chose to include this in favor of our `<InfiniteList />` implementation:

- There is clearly something buggy with `<InfiniteList />` when used in the context of a non-window scrollable element. I believe this is related to calculations of placeholder blocks not correctly considering offset from the window for these elements. That the media library works without issue seems to be a fluke. After spending several hours debugging this issue to no avail, I found [`react-virtualized`](https://github.com/bvaughn/react-virtualized), which is [well-tested](https://codecov.io/github/bvaughn/react-virtualized), [thoroughly documented](https://github.com/bvaughn/react-virtualized/tree/master/docs), highly configurable, [actively maintained](https://github.com/bvaughn/react-virtualized/commits/master), [popular](https://github.com/bvaughn/react-virtualized/stargazers), and [issues are promptly resolved](https://github.com/bvaughn/react-virtualized/issues).
- `react-virtualized` has support for all of the features I could surmise us needing, and additionally supports the render callback which I had anticipated needing to implement in `<InfiniteList />`, required for proper re-querying of restored pagination state. More detail at https://github.com/Automattic/wp-calypso/pull/5135#issuecomment-221672980, the support being [`<VirtualScroll />`'s `onRowsRendered` callback](https://github.com/bvaughn/react-virtualized/blob/master/docs/VirtualScroll.md).

/cc @blowery @mtias for feedback on consideration of this alternative to infinite list

__Testing instructions:__

Repeat the testing instructions from #3197, with particular care for:
- Searching
- Resetting search
- Compact view
- Infinite scrolling

timmyc commented:
I was just looking at trying to make infinite scroll work for the term selector as well, but due to the varied height of a row with children, it was not going to work.  Looks like VirtualScroll could possibly be a good fit there as well.

timmyc commented on client/my-sites/post-selector/selector.jsx:
```diff
@@ -97,136 +109,202 @@ const PostSelectorPosts = React.createClass( {
 
-	renderPlaceholder() {
-		return ( <ul>{ this.renderPlaceholderItem() }</ul> );
+	getRowCount() {
```
So this row count excludes child `items` correct?  But this isn't an issue since the `getItemHeight` factors them in to the height correct?

aduth replied:
>So this row count excludes child `items` correct?  But this isn't an issue since the `getItemHeight` factors them in to the height correct?

Correct on both accounts.

timmyc commented on client/my-sites/post-selector/selector.jsx:
```diff
@@ -97,136 +109,202 @@ const PostSelectorPosts = React.createClass( {
-				'is-type-labels-visible': showTypeLabels
-			}
+	renderItem( item ) {
+		const postType = get( this.props.postTypes, [ item.type, 'labels', 'singular_name' ], '' );
```
Although this is probably a fairly quick lookup, but would it be worth conditionally getting the postType based off of `isTypeLabelsVisible`?

aduth replied:
>Although this is probably a fairly quick lookup, but would it be worth conditionally getting the postType based off of `isTypeLabelsVisible`?

Inlined all of the type label rendering conditionally in 62431fd85ade2e6386ed664e4ed03d8eca113b63 (prevents type label from being rendered at all when `isTypeLabelsVisible` is `false`).

timmyc commented:
Took this for a test drive, and the Infinite Scroll is quite smooth.  I kept a close eye on the network tab and all looked well, even with paginated search.  I did however experience the issue in the screen grab below a few times.  When typing a bit quickly in the search box, the component gets stuck in a permanent loading state, even though you can see in the network tab that the request has completed successfully.

![inifinite](https://cloud.githubusercontent.com/assets/22080/15730351/b4bda350-2820-11e6-948e-528f1878526d.gif)

I would love to utilize `react-virtualized` in #5366 too!

timmyc commented:
Also it looks like the padding still isn't quite right when no results are found:

<img width="370" alt="edit_post_ _testingtimmy2_wordpress_com_ _wordpress_com" src="https://cloud.githubusercontent.com/assets/22080/15730381/11ed7438-2821-11e6-8eda-0506b7fbd6ee.png">

aduth commented on assets/stylesheets/_components.scss:
```diff
@@ -370,3 +370,12 @@
 @import 'support/support-user/style';
 @import 'vip/vip-logs/style';
+
+// ==========================================================================
```
Thoughts on declaring external dependency imports here? Considered creating a separate `_vendor.scss` file instead.

Also a bit unsure about reaching into `node_modules` as I've done here, but it seems to work fine.

timmyc replied:
Hmm that is a great question.  Seems fine to me to place the import here, but have to agree grabbing from `node_modules` looks/feels strange.  The [file](https://github.com/bvaughn/react-virtualized/blob/master/styles.css) isn't huge, but we could take the approach done with `component/tip` and grab the base styles and place within calypso itself [#](https://github.com/Automattic/wp-calypso/blob/dabf92fa84865576ecd8d806bd21c9cc4184e73d/client/components/popover/style.scss)

aduth replied:
>Hmm that is a great question.  Seems fine to me to place the import here, but have to agree grabbing from `node_modules` looks/feels strange.  The [file](https://github.com/bvaughn/react-virtualized/blob/master/styles.css) isn't huge, but we could take the approach done with `component/tip` and grab the base styles and place within calypso itself [#](https://github.com/Automattic/wp-calypso/blob/dabf92fa84865576ecd8d806bd21c9cc4184e73d/client/components/popover/style.scss)

Yeah, not a straight-forward answer here. The issue with copying the styles is more a maintainability concern, requiring us to keep on top of syncing the styles any time we choose to upgrade `react-virtualized`. As strange as it looks to reach into `node_modules`, I can't think of a case where it wouldn't work as we'd expect, especially since [`node_modules` Makefile target is a prerequisite to our style build](https://github.com/Automattic/wp-calypso/blob/3901c935163c986626ca7c73e0ad2bdc72350ecf/Makefile#L129).

aduth commented:
@timmyc : Thanks for taking this for a spin! As always, you amaze me with the edge cases you're able to uncover 😄 

>I did however experience the issue in the screen grab below a few times.

What I determined here was that this would occur when the search would result in a single result, since it seemed that ReactVirtualized wouldn't cause the list to re-render if it determined that the row count hadn't changed. Since the row count when loading (1 = placeholder) was the same, it didn't re-render after the posts response was received. I resolved this by causing a forced render any time the component changes from loading to not loading (previously, was similar to this but limited to new posts set qualifying as either "no posts" or "search produced no results" cases).

>Also it looks like the padding still isn't quite right when no results are found:

There was 2px additional height for the list items when results exist, which I [recreated by adding more padding to the empty label](https://github.com/Automattic/wp-calypso/blob/62431fd85ade2e6386ed664e4ed03d8eca113b63/client/my-sites/post-selector/style.scss#L63-L65) in the rebased 62431fd. 

![After](https://cloud.githubusercontent.com/assets/1779930/15833025/baf68b44-2bf2-11e6-9726-7b02f4fff4d3.png)

timmyc commented:
> What I determined here was that this would occur when the search would result in a single result

Nice find - I tried to figure out what was going on for a bit but was puzzled!  I can confirm this is now fixed - nice work.  No results text is looking much better now as well.

I really like how this component feels while scrolling - it feels smoother to me, quicker almost.  Not sure about the proper solution for the styles inclusion, but left my thoughts on that above.

I'm good with this PR, but not sure if @mtias or @blowery want to take a test drive also.  I would definitely use this approach with the `term-selector` as my home-grown solution is not nearly as elegant as this for IS with nested results.

blowery commented:
<img width="411" alt="edit_post_ _b19y_ _wordpress_com" src="https://cloud.githubusercontent.com/assets/14350/15837416/2e08bcf6-2c07-11e6-910f-f91f38afb5ca.png">

some odd wrapping with long post titles..

aduth commented:
>some odd wrapping with long post titles..

Ah, good catch. This is likely due to the naive calculation of item height using a fixed value per item. We'll either need to not allow titles to wrap or update the calculation to use the rendered height.

aduth commented:
@blowery : Should be resolved as of 36e021f.

![Computed row height](https://cloud.githubusercontent.com/assets/1779930/15840042/3c6d6784-2c15-11e6-93c0-a507e243361a.png)

aduth commented:
Rebased to resolve conflicts introduced with merging of #5135.

On note of external dependency stylesheet, while I wish there was an obvious and easy way to namespace the components, I'm pretty content with how we're importing it now. That said, I moved the import out of `_components.scss` and into the main `styles.scss`.

In 2b017eb, I changed to import the components directly from their source file instead of the root export, and added an alias to the webpack configuration (to avoid `/dist/commonjs` being part of the import path).

timmyc commented:
@aduth Just took this for another spin and it is working quite nicely - even with really long titles and deeply nested pages.  I think we should get this out there and put `react-virtualized` through the paces.  Marking as ready to merge but if @blowery or anyone else would like one last pass please feel free to do so.

alisterscott commented:
Great work on this, I really like it.
